File: frontend/src/services/notificationService.ts
import { get, post, put, del } from "./api";
import type { Notification, PaginatedResponse, PaginationParams } from "../types";

function qs(params: PaginationParams): Record<string, string | number | boolean | undefined> {
  return {
    page: params.page,
    per_page: params.per_page,
    search: params.search,
    sort: params.sort,
    order: params.order,
  };
}

export const notificationService = {
  list(params: PaginationParams = {}): Promise<PaginatedResponse<Notification>> {
    return get<PaginatedResponse<Notification>>("/notifications", qs(params));
  },

  get(id: number): Promise<Notification> {
    return get<Notification>(`/notifications/${id}`);
  },

  create(data: Partial<Notification>): Promise<Notification> {
    return post<Notification>("/notifications", data);
  },

  update(id: number, data: Partial<Notification>): Promise<Notification> {
    return put<Notification>(`/notifications/${id}`, data);
  },

  delete(id: number): Promise<void> {
    return del<void>(`/notifications/${id}`);
  },

  nonLues(): Promise<Notification[]> {
    return get<Notification[]>("/notifications/non-lues");
  },

  lire(id: number): Promise<Notification> {
    return post<Notification>(`/notifications/${id}/lire`);
  },

  lireToutes(): Promise<{ message: string }> {
    return post<{ message: string }>("/notifications/lire-toutes");
  },
};

File: gs_operateur_mobile/lib/services/notification_service.dart
import '../models/notification.dart';
import 'api_service.dart';

class NotificationService {
  final ApiService _api = ApiService();

  Future<List<Notification>> list({String? search, int page = 1, int perPage = 20}) async {
    final response = await _api.get('/notifications', queryParameters: {
      'page': page,
      'per_page': perPage,
      if (search != null && search.isNotEmpty) 'search': search,
    });

    if (response['success'] == true) {
      final data = response['data'];
      return (data is List ? data : data['data'] ?? [])
          .map<Notification>((n) => Notification.fromJson(n))
          .toList();
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<List<Notification>> nonLues() async {
    final response = await _api.get('/notifications/non-lues');
    if (response['success'] == true) {
      final data = response['data'];
      return (data is List ? data : [])
          .map<Notification>((n) => Notification.fromJson(n))
          .toList();
    }
    throw Exception(response['message'] ?? 'Erreur');
  }

  Future<void> marquerCommeLue(int id) async {
    await _api.patch('/notifications/$id/lire');
  }

  Future<void> marquerToutesCommeLues() async {
    await _api.patch('/notifications/lire-toutes');
  }

  Future<void> delete(int id) async {
    await _api.delete('/notifications/$id');
  }

  Future<int> countNonLues() async {
    try {
      final items = await nonLues();
      return items.length;
    } catch (_) {
      return 0;
    }
  }
}
